### ISSUE TITLE
Let QApplication remove the command line arguments it recognizes before GDCommandLine parses them

### ISSUE THREAD

**Issue by vedgy** [goldendict/goldendict#1623]
### Steps to reproduce:
1. Run `goldendict -style Fusion` from terminal.
### Actual results:
1. GoldenDict is styled as expected.
2. The translation for "fusion" is displayed.
### Expected results:
1. GoldenDict is styled as expected.
2. GoldenDict's Welcome page is displayed.

From https://doc.qt.io/qt-5/qapplication.html#QApplication (similar documentation for Qt 4's `QApplication`):
> Note: argc and argv might be changed as Qt removes command line arguments that it recognizes.

This issue is not trivially fixable, because GoldenDict creates a `QApplication` if `GDCommandLine` encounters the `--show-error-file` command line option and a `QHotkeyApplication` otherwise. One possible fix is to parse the special `--show-error-file` command line option separately from the other options. This option is passed by the previously crashed GoldenDict issue itself in `termHandler()`. So an additional special-purpose simplified command line parser can support only this option and the filename that follows it; early-return if it encounters some other command line option in place of `--show-error-file`.

**Comment by vedgy** [comment]
> It popup a message box that does absolutely nothing, then quit with 0 as the exit code.

Have you tested the feature or are just guessing, because someone has already removed the feature from goldendict-ng? I just added the line `throw std::runtime_error("jumpy");` at the beginning of `ArticleView::jumpToTargetArticle()` and got this message box, as intended by the author I believe:
![show-error-file](https://user-images.githubusercontent.com/5929075/234888210-e8286ef6-6828-4642-8e19-0716b147063c.png)

**Comment by shenlebantongying** [comment]
The termination handler does not work on Windows in the release build.

Can you test what's will on this screen if `-O2` is enabled?

Can u measure the performance impact of `-rdynamic`? 

The root reason for this program needing a custom termination handler in `termination.cc` comes from the laziness of using `DEF_EX` macro which just defines an empty exception with nothing else. They are usually not handled at all. A lot of them are used in some dict formats. The code just defines an empty one, then doesn't do anything about them.

Just replace the lazy `DEF_EX` and include at least the dictionary's name and a message about exactly why something failed, then this termination handler's usefulness will be almost gone.

When a dict parsing goes wrong, the app doesn't even need to quit/crash, just mark it as dirty.

**Comment by nonwill** [comment]
> Have you tested the feature or are just guessing, because someone has already removed the feature from goldendict-ng?

@shenlebantongying is just doing similar things again.
I can't make any comments on #1624, but the implementation has obvious problems.:
![2023-04-27_225311](https://user-images.githubusercontent.com/46510529/234902337-c88ca1d9-3883-45dd-bdac-c7c74d8c61b1.png)

Word “Http://” exists in wikipedia_en_simple_all_maxi_2021-09.zim, would the processing here cause similar queries to have no results returned?

But at the end of content posted on #1624 shenlebantongying emphasizes:
> And it works 🦄 🔥 💯 🥇

I guess it's just shenlebantongying’s guess too.

**Comment by vedgy** [comment]
> The termination handler does not work on Windows in the release build.

Are you testing a (semi-)official GoldenDict build? How do you reproduce a thrown exception?

> Can you test what's will on this screen if `-O2` is enabled?

Got practically the same backtrace in a Release build (-O2). 
 
> Can u measure the performance impact of `-rdynamic`?

Judging by [these](https://stackoverflow.com/questions/12634114/will-adding-the-rdynamic-linker-option-to-gcc-g-impact-performance) [answers](https://stackoverflow.com/questions/20491130/impact-disadvantages-of-rdynamic-gcc-option), the flag shouldn't noticeably affect performance of the relatively small GoldenDict application.

> The primiary reason for this program needing a custom termination handler in `termination.cc` comes from the laziness of using `DEF_EX` macro which just defines an empty exception with nothing else. They are usually not handled at all. A lot of them are used in some dict formats. The code just defines an empty one, then doesn't do anything about them.
> 
> Just avoid the lazy `DEF_EX` and include at least the dictionary's name and a message about exactly why something failed, then this termination handler's usefulness will be almost gone.
> 
> Fully defining the `Dictionary::Ex` with dict's ID/names, and probably other empty exceptions will give more useful info.

I don't suppose the strings would provide more information than a backtrace. Plus showing a message box in case of a crash is more user-friendly than crashing silently. With the message box, end users are more likely to report an uncaught exception in a way that would allow to understand and fix the bug.

> When a dict parsing goes wrong, the app doesn't even need to quit/crash, just mark it as dirty.

Quiet wrong parsing is not necessarily better. A crash with a backtrace allows to learn about and fix the parsing bug.

**Comment by shenlebantongying** [comment]
> more detailed source file and line location, but no call stack

Yes, we probably can redefine `Dictionary::Ex` and possibly a few others to include more info like `__FILE__` and `__LINE__` so that when the termination handler receives those types of exceptions, we will know exactly where.

However, this needs massive changes.

**Comment by vedgy** [comment]
> > more detailed source file and line location, but no call stack
> 
> Yes, we probably can redefine `Dictionary::Ex` and possibly a few others with more details like `__FILE__` and `__LINE__` so that when the termination handler receives those types of exceptions, we will know exactly where.

My point is that even if all exception messages are unique and therefore allow to uniquely identify the file and line in code, that still wouldn't provide a call stack, unlike the feature you suggest is almost useless.

**Comment by xiaoyifang** [comment]
https://github.com/goldendict/goldendict/blob/192f0d9f25350c42ff7c536d49fefdc429202607/main.cc#L304-L316

IMO, If the error is important ,it should never be removed.  If not important, it should not exist.

**Comment by vedgy** [comment]
> IMO, If the error is important ,it should never be removed. If not important, it should not exist.

The error message is displayed to the user and the user is given a chance to copy the backtrace either from the dialog or from the temporary file, until the dialog is dismissed. The dialog message tells that much.

On the other hand, removing the temporary file is not necessary, because it is small enough and is removed automatically when the system temporary directory is cleared on system reboot.

I think both removing and not removing the file is reasonable. If either approach causes problems, one can easily switch to the other one. But not if the feature is removed entirely, of course.

**Comment by xiaoyifang** [comment]
errFileName  is passed through the arguments ，I do not think it is a temporary file.
remove the file is a destructive action.

**Comment by vedgy** [comment]
`--show-error-file` is a hidden undocumented command line option. It is reserved solely for GoldenDict's internal use:
https://github.com/goldendict/goldendict/blob/192f0d9f25350c42ff7c536d49fefdc429202607/termination.cc#L144-L160

**Comment by xiaoyifang** [comment]
https://github.com/goldendict/goldendict/blob/192f0d9f25350c42ff7c536d49fefdc429202607/main.cc#L175-L182

**Comment by vedgy** [comment]
> And it will expose a security risk to the system.

What's the risk? That a file at the path passed via the command line is removed?

**Comment by xiaoyifang** [comment]
![image](https://user-images.githubusercontent.com/105986/236652691-45b0d4b8-8856-4613-92d3-0a9e897ce2c5.png)

**Comment by vedgy** [comment]
Does the `rm` program (and its equivalent command on Windows) make a system it is installed on vulnerable too? If not, what's the difference? Do you expose unsanitized access to GoldenDict command line over the Internet?

**Comment by xiaoyifang** [comment]
the users who use rm know what they are doing .
I doubt this can be applied to goldendict --show-error-file

**Comment by vedgy** [comment]
> the users who use rm know what they are doing . I doubt this can be applied to goldendict --show-error-file IMO ,it is a bad design in GoldenDict .

So the big and scary risk is that a user **accidentally** passes two arguments to GoldenDict: `--show-error-file` and a path to an existing file; GoldenDict shows the file's contents, warns the user that the file at this path is about to be removed, the user doesn't notice or ignores the message, closes the dialog and then regrets that the file has been removed. And this constitutes "a bad design" of GoldenDict.

I am sure many programs have a design just as "bad". I definitely disagree about this one and don't see any benefit in further discussing the matter.
